--- a/net/minecraft/world/level/block/entity/SculkCatalystBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/SculkCatalystBlockEntity.java
@@ -34,7 +_,9 @@
    }
 
    public static void m_222779_(Level p_222780_, BlockPos p_222781_, BlockState p_222782_, SculkCatalystBlockEntity p_222783_) {
+      org.bukkit.craftbukkit.event.CraftEventFactory.sourceBlockOverride = p_222783_.m_58899_(); // CraftBukkit - SPIGOT-7068: Add source block override, not the most elegant way but better than passing down a BlockPosition up to five methods deep.
       p_222783_.f_279609_.m_280490_().m_222255_(p_222780_, p_222781_, p_222780_.m_213780_(), true);
+      org.bukkit.craftbukkit.event.CraftEventFactory.sourceBlockOverride = null; // CraftBukkit
    }
 
    public void m_142466_(CompoundTag p_222787_) {
