--- a/net/minecraft/world/entity/npc/AbstractVillager.java
+++ b/net/minecraft/world/entity/npc/AbstractVillager.java
@@ -34,8 +_,19 @@
 import net.minecraft.world.level.ServerLevelAccessor;
 import net.minecraft.world.level.pathfinder.BlockPathTypes;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.inventory.CraftMerchant;
 
 public abstract class AbstractVillager extends AgeableMob implements InventoryCarrier, Npc, Merchant {
+
+   // CraftBukkit start
+   private CraftMerchant craftMerchant;
+
+   @Override
+   public CraftMerchant getCraftMerchant() {
+      return (craftMerchant == null) ? craftMerchant = new CraftMerchant(this) : craftMerchant;
+   }
+   // CraftBukkit end
+
    private static final EntityDataAccessor<Integer> f_35262_ = SynchedEntityData.m_135353_(AbstractVillager.class, EntityDataSerializers.f_135028_);
    public static final int f_149991_ = 300;
    private static final int f_149992_ = 8;
@@ -116,6 +_,7 @@
          CriteriaTriggers.f_10585_.m_70959_((ServerPlayer)this.f_35263_, this, p_35274_.m_45368_());
       }
 
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.player.TradeWithVillagerEvent(this.f_35263_, p_35274_, this));
    }
 
    protected abstract void m_8058_(MerchantOffer p_35299_);
@@ -164,9 +_,9 @@
    }
 
    @Nullable
-   public Entity m_5489_(ServerLevel p_35295_) {
+   public Entity changeDimension(ServerLevel p_35295_, net.minecraftforge.common.util.ITeleporter teleporter) {
       this.m_7996_();
-      return super.m_5489_(p_35295_);
+      return super.changeDimension(p_35295_, teleporter);
    }
 
    protected void m_7996_() {
