--- a/net/minecraft/world/item/LeadItem.java
+++ b/net/minecraft/world/item/LeadItem.java
@@ -2,6 +_,7 @@
 
 import net.minecraft.core.BlockPos;
 import net.minecraft.tags.BlockTags;
+import net.minecraft.world.InteractionHand;
 import net.minecraft.world.InteractionResult;
 import net.minecraft.world.entity.Mob;
 import net.minecraft.world.entity.decoration.LeashFenceKnotEntity;
@@ -11,6 +_,8 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.phys.AABB;
+import org.bukkit.craftbukkit.CraftEquipmentSlot;
+import org.bukkit.event.hanging.HangingPlaceEvent;
 
 public class LeadItem extends Item {
    public LeadItem(Item.Properties p_42828_) {
@@ -24,6 +_,7 @@
       if (blockstate.m_204336_(BlockTags.f_13039_)) {
          Player player = p_42834_.m_43723_();
          if (!level.f_46443_ && player != null) {
+            prepareBindPlayerMobs(p_42834_.m_43724_()); // CraftBukkit - Pass hand
             m_42829_(player, level, blockpos);
          }
 
@@ -33,7 +_,14 @@
       }
    }
 
-   public static InteractionResult m_42829_(Player p_42830_, Level p_42831_, BlockPos p_42832_) {
+   // Magma start - Move InteractionHand to different constructor.
+   private static net.minecraft.world.InteractionHand enumhand = null;
+   public static void prepareBindPlayerMobs(net.minecraft.world.InteractionHand enumhan) {
+      LeadItem.enumhand = enumhand;
+   }
+   // Magma end
+
+   public static InteractionResult m_42829_(Player p_42830_, Level p_42831_, BlockPos p_42832_) { // CraftBukkit - Add EnumHand
       LeashFenceKnotEntity leashfenceknotentity = null;
       boolean flag = false;
       double d0 = 7.0D;
@@ -45,12 +_,31 @@
          if (mob.m_21524_() == p_42830_) {
             if (leashfenceknotentity == null) {
                leashfenceknotentity = LeashFenceKnotEntity.m_31844_(p_42831_, p_42832_);
+
+               // CraftBukkit start - fire HangingPlaceEvent
+               org.bukkit.inventory.EquipmentSlot hand = CraftEquipmentSlot.getHand(enumhand);
+               HangingPlaceEvent event = new HangingPlaceEvent((org.bukkit.entity.Hanging) leashfenceknotentity.getBukkitEntity(), p_42830_ != null ? (org.bukkit.entity.Player) p_42830_.getBukkitEntity() : null, p_42831_.getWorld().getBlockAt(i, j, k), org.bukkit.block.BlockFace.SELF, hand);
+               p_42831_.getCraftServer().getPluginManager().callEvent(event);
+               if (event.isCancelled()) {
+                  leashfenceknotentity.m_146870_();
+                  return InteractionResult.PASS;
+               }
+               // CraftBukkit end
+
                leashfenceknotentity.m_7084_();
             }
 
+            // CraftBukkit start
+            if (org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerLeashEntityEvent(mob, leashfenceknotentity, p_42830_, enumhand).isCancelled()) {
+               continue;
+            }
+            // CraftBukkit end
+
             mob.m_21463_(leashfenceknotentity, true);
             flag = true;
          }
+
+         enumhand = null;
       }
 
       if (flag) {
