--- a/net/minecraft/world/inventory/FurnaceResultSlot.java
+++ b/net/minecraft/world/inventory/FurnaceResultSlot.java
@@ -43,10 +_,12 @@
       if (player instanceof ServerPlayer serverplayer) {
          Container container = this.f_40218_;
          if (container instanceof AbstractFurnaceBlockEntity abstractfurnaceblockentity) {
+            abstractfurnaceblockentity.prepareAwardUsedRecipesAndPopExperience(p_39558_, this.f_39540_); // CraftBukkit // Magma
             abstractfurnaceblockentity.m_155003_(serverplayer);
          }
       }
 
       this.f_39540_ = 0;
+      net.minecraftforge.event.ForgeEventFactory.firePlayerSmeltedEvent(this.f_39539_, p_39558_);
    }
 }
