--- a/net/minecraft/world/level/chunk/LevelChunkSection.java
+++ b/net/minecraft/world/level/chunk/LevelChunkSection.java
@@ -178,6 +_,12 @@
       return this.f_187995_.m_63087_(p_204434_, p_204435_, p_204436_);
    }
 
+   // CraftBukkit start
+   public void setBiome(int i, int j, int k, Holder<Biome> biome) {
+      this.f_187995_.set(i, j, k, biome);
+   }
+   // CraftBukkit end
+
    public void m_280631_(BiomeResolver p_282075_, Climate.Sampler p_283084_, int p_282310_, int p_281510_, int p_283057_) {
       PalettedContainer<Holder<Biome>> palettedcontainer = this.f_187995_.m_238334_();
       int i = 4;
